### PR TITLE
:zap: Changed Catmod 'fs' to use app data directories for Linux, Windows, macOS

### LINKED ISSUE #224: [catmod fs] Bring in-line with XDG Base Directory Specification
**Describe the current state of the problem**
Currently, by default, the "Desktop's file system" catmod uses `~/${author_name}/${project_name}` which has the potential to heavily pollute the user's home directory when ct.js becomes more popular.

**Describe the solution you'd like**
Instead it should follow the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/latest/), which states the following locations should be used for various purposes.

`$XDG_DATA_HOME` as the base directory for [data files](https://en.wikipedia.org/wiki/Data_file). Defaults to `~/.local/share`
`$XDG_CONFIG_HOME` as the location for [configuration files](https://en.wikipedia.org/wiki/Configuration_file). Defaults to `~/.config`

This part is more subjective, but I do not think the author's name should be used in the path. Instead a separate "Organization Name" should be created and used in it's place. It's conceivable that in some organizations, there is not one specific author, but for indie development, I believe a lot of people would have the two be separate.

For an example. If I published a game called Vroom under a Organization called Ursid I would expect the data and config directories to likely be as follows:
`~/.local/share/ursid/vroom/`
`~/.config/ursid/vroom/`

**Describe alternatives you've considered (optional)**
I do not think that there is any good alternative to using the specification, however I can see the separation of author and organization being rejected. However if it does, I would suggest renaming it to "developer name".

**Additional context**
I have no idea if MacOS follows the specification and I would be surprised if Windows did. This is mainly for Linux.

### PR DESCRIPTION
**Pull request by pennro-se**
Closes #224.

**Changes proposed in this pull request:**
- Follow the XDG Base Directory Specification For Linux
- Use AppData for Windows
- Use Library/Application Support for macOS
- Updated Catmod 'fs' types
- Updated Catmod 'fs' documentation

**Help wanted:** (optional, e.g. a request for GUI revamp)
- Should projects in development be tagged in a way to stop users from overwriting the app data of their exported games?
- Putting the Operating System and Browser info in the ct core would help keep things DRY (reused function from Catmod 'keyboard polyfill')

Info on AppData locations collected [here](https://knowledge.ursid.ca/books/misc/page/application-data-and-where-to-find-it) from Resources listed.
**Ping @CosmoMyzrailGorynych**

### CODE REVIEW

File: app/data/ct.libs/fs/index.js [hunk 1]
```diff
@@ -96,6 +137,9 @@ try {
 
     ct.fs.rename = ct.fs.move;
     ct.fs.exists = ct.fs.stat;
+
+    // Util
+    function contains(s, ss) { return String(s).indexOf(ss) !== -1; }
```
Comment (amsam0) [inline app/data/ct.libs/fs/index.js]: Why is this here? Doesn't seem to do anything 🤔

File: app/data/ct.libs/fs/index.js [hunk 2]
```diff
@@ -2,21 +2,59 @@
...
+    const getAppData = (home, operatingSystem) => {
+        switch(operatingSystem) {
+            case operatingSystems.Windows:
+                return process.env.AppData;
+            case operatingSystems.macOS:
+                return `${home}/Library/Application Support`;
+            case operatingSystems.Linux:
+                return process.env.XDG_DATA_HOME || `${home}/.local/share`;
+            // Don't know what to do for ChromeOS or iOS, do they use AppData or
+            // Should those default to LocalStorage?
```
Comment (amsam0) [inline app/data/ct.libs/fs/index.js:56]: It's unlikely to be able to write to mobile devices (so far only android has been possible (and thats not merged yet)), so I'd say don't do anything.

File: app/data/ct.libs/fs/README.md [hunk 3]
```diff
@@ -4,7 +4,7 @@ It allows you to easily save and load JSON objects, as well as plain text data.
...
 
-By default, all the file operations will be performed relative to a special directory created for your game, which is templated as a `${User's home directory}/${Author name from the settings tab}/${Project's name from the same tab}`, e.g. `/home/comigo/Cosmo Myzrail Gorynych/Platformer complete tutorial` on a Linux machine. You can inspect this behavior by calling `ct.fs.getPath(...)`, or simply by reading ct.fs.gameFolder parameter.
+By default, all the file operations will be performed relative to a special directory created for your game, which is templated as a `${User Application Data}/${Author name from the settings tab}/${Project's name from the same tab}`, e.g. `/home/comigo/.local/share/Cosmo Myzrail Gorynych/Platformer complete tutorial` on a Linux machine. You can inspect this behavior by calling `ct.fs.getPath(...)`, or simply by reading ct.fs.gameFolder parameter.
```
Comment (amsam0) [inline app/data/ct.libs/fs/README.md]: Maybe replace `comigo` with `<user>`? Someone might get confused

**Review by amsam0 — CHANGES_REQUESTED** [review]
Looks good!

File: app/data/ct.libs/fs/index.js (hunk 1 above)
Comment (pennro-se) [reply app/data/ct.libs/fs/index.js]: It's used earlier in the file, to check for substrings in `navigator.platform`. I borrowed the code from `keyboard.polyfill`

File: app/data/ct.libs/fs/index.js (hunk 2 above)
Comment (pennro-se) [reply app/data/ct.libs/fs/index.js:56]: At the moment, I have it defaulting back to using bare `home` if it's not Linux, Windows, or macOS.

File: app/data/ct.libs/fs/index.js (hunk 1 above)
Comment (amsam0) [reply app/data/ct.libs/fs/index.js]: maybe it should be farther up?

File: app/data/ct.libs/fs/README.md (hunk 3 above)
Comment (pennro-se) [reply app/data/ct.libs/fs/README.md]: I'll see about rewriting the documentation to be less [magical number-y](https://en.wikipedia.org/wiki/Magic_number_%28programming%29#Unnamed_numerical_constants).

File: app/data/ct.libs/fs/index.js (hunk 1 above)
Comment (pennro-se) [reply app/data/ct.libs/fs/index.js]: Can do

**Review by amsam0 — APPROVED** [review]

**Comment by CosmoMyzrailGorynych** [comment]
I will standardize formatting a bit, but in overall, nice work!

**Comment by CosmoMyzrailGorynych** [comment]
@JulianWebb 

> Should projects in development be tagged in a way to stop users from overwriting the app data of their exported games?

I suppose no; if you need it, use `ct.fs.gameFolder`.

> Putting the Operating System and Browser info in the ct core would help keep things DRY (reused function from Catmod 'keyboard polyfill')

I feel that OS detection should be in the `ct.desktop` module. Though this module is currently nothing but one function :D In any way, OS detection should go in a separate issue.
Checking for browsers is a bad practice (you should check against available functionality, not browsers, as the latter are always changing and any hack quickly becomes outdated).
